rfcs/20210510-binary-size-reduction/README.md

```diff
@@ -0,0 +1,156 @@
+includes necessary code and it includes some unnecessary, too. Unnecessary code
+can be either of:
+* Instantiations of templated classes.
+* Specializations of templated classes or class members.
+* Non-templated class methods definitions.
+* ... (maybe something else we don't experience in the library) ...
+
+This behavior can't be changed or influenced externally.
+
+## Implementation Details
+
+Main mechanism:
+
+```cmake
+# options.cmake:
+
+foreach(impl in ${DNNL_IMPLS})
+    foreach(prop in ${DNNL_PROPS})
+        string(TOUPPER ${impl} uimpl)
+        string(TOUPPER ${prop} uprop)
+        ...
+        set(BUILD_${uimpl}_${uprop} TRUE)
+    endforeach()
+endforeach()
+```
+
+```cpp
+// src/common/impl_registration.hpp (new header):
+
+#if BUILD_CONVOLUTION_FWD
+#define REGISTER_CONVOLUTION_FWD(...) __VA_ARGS__
+#else
+#define REGISTER_CONVOLUTION_FWD(...)
+#endif
+
+#if BUILD_CONVOLUTION_BWD
+#define REGISTER_CONVOLUTION_BWD(...) __VA_ARGS__
+#else
+#define REGISTER_CONVOLUTION_BWD(...)
+#endif
+```
+
+```cpp
+// src/cpu/cpu_convolution_list.cpp:
+const std::map<conv_impl_key_t, std::vector<pd_create_f>> impl_list_map {
+    // FWD fp
+    {{forward, f32, f32, f32}, {
+        REGISTER_CONVOLUTION_FWD(CPU_INSTANCE_X64(ip_convolution_fwd_t))
+        ...
+        REGISTER_CONVOLUTION_FWD(CPU_INSTANCE(ref_fused_convolution_fwd_t))
+        nullptr,
+    }},
+```
+
+Additionally, to overcome some of [2], instantiations in separate translations
+units should be wrapped as well:
+
+```cpp
+// src/cpu/ref_convolution.cpp:
+
+REGISTER_CONVOLUTION_FWD(template struct ref_convolution_fwd_t<f32>);
+...
+```
+
+To resolve unnecessary code inclusions, it is enough to separate, i.e. for
+propagation kind, forward implementation details from backward into different
+translation units. Then linker will not include anything from backward if it was
+switched off by user. This is likely not the cheapest but straightforward
+solution which allows to leave the code written itself intact.
+
+## Open questions
+* GEMM auto-gen kernel take a lot of disk space and the plan is to replace them
+  with correspondent BRGEMM implementation. The problem is the final transition
+  may happen in 8+ month which is a long period. There is an option to disable
+  those in compile time as well, but this might hit performance. If new level
+  of performance without those kernels is acceptable, we may add a switch to
+  remove them in compile time as well alongside other primitives. Should we
+  consider doing it?
+
+* Is it reasonable to enable data type option at all?
+  - Likely oneDNN team needs to experiment with shrinking data type templates
+    and measure the impact on binary size firstly to have a proper answer.
+
+* If enable ISA as an option:
+  - What does it mean to enable current ISA - use implementations which support
+    only this ISA, or follow some comparison system and include all up to
+    selected?
```

Comment:
Hi @dzarukin 
I was wondering if with `DDNNL_ARCH_OPT_FLAGS=mavx2` (say) do we not generate jit`ed code for lower ISA. I was thinking of using this flag as another mean to reduce the library size  [Though I don't know how to deal with the scenario where a particular shape is handled only by the lower ISA jit kernel]

Reply:
Hi @alokbakshi, I'm not sure this is the same. Using this flag doesn't prevent user from running the library on higher ISA and still get optimized jit-ted code for it. This flag instructs compiler to generate avx2 (say) optimized **C/C++** code for simple and reference implementations, so in general I don't think this is applicable here.

Reply:
Hi @dzarukin 

But it will prevent user from running the library on _lower_ ISA, NO?. Say we simply don't generate any SSE jit kernels with `DDNNL_ARCH_OPT_FLAGS=mavx2` flag thereby reducing the library size. I think this approach is complementary to what you've suggested -- namely, we don't generate jit kernels whose ISA requirements is more than the current ISA.

Reply:
OKey, I think I got what you are trying to say here...
I admit that user unlikely will run an application compiled with mavx2 on sse4.1 due to illegal instruction likely popping up. The problem here is that relying on ARCH_OPT_FLAG can't be done since there are values like `native` which can't be treated as certain ISA without extra hacks. And additional dependency may lead to unpleasant side effects, so I would refrain from relying on anything when it comes to ISA. Would you agree with this?

Reply:
Hi @dzarukin 

Yes that makes sense `--march=native` flags seem specially bothersome